dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/filters/DatacenterIdFilter.java

```diff
@@ -17,31 +17,51 @@
 package org.jclouds.dimensiondata.cloudcontrol.filters;
 
 import com.google.common.base.Supplier;
+import org.jclouds.compute.reference.ComputeServiceConstants;
 import org.jclouds.http.HttpException;
 import org.jclouds.http.HttpRequest;
 import org.jclouds.http.HttpRequestFilter;
+import org.jclouds.location.Region;
 import org.jclouds.location.Zone;
+import org.jclouds.logging.Logger;
 
+import javax.annotation.Resource;
 import javax.inject.Inject;
+import javax.inject.Named;
+import java.util.Map;
 import java.util.Set;
 
+import static com.google.common.base.Preconditions.checkNotNull;
+
 /**
- * Adds set of Datacenter IDs as set in jclouds.zones JVM property.
+ * Adds set of Datacenter IDs using configuration.
  */
 public class DatacenterIdFilter implements HttpRequestFilter {
 
-   private final Supplier<Set<String>> datacenterIdsSupplier;
+   @Resource
+   @Named(ComputeServiceConstants.COMPUTE_LOGGER)
+   private Logger logger = Logger.NULL;
+
+   protected final Set<String> datacenterIds;
+
+   protected String queryParam = "datacenterId";
 
    @Inject
-   DatacenterIdFilter(@Zone Supplier<Set<String>> datacenterIdsSupplier) {
-      this.datacenterIdsSupplier = datacenterIdsSupplier;
+   DatacenterIdFilter(@Region Supplier<String> implicitRegionIdSupplier,
+         @Zone Supplier<Map<String, Supplier<Set<String>>>> regionIdToZoneIdsSupplier) {
+      checkNotNull(implicitRegionIdSupplier, "implicitRegionIdSupplier");
+      checkNotNull(regionIdToZoneIdsSupplier, "regionIdToZoneIdsSupplier");
+      final String regionId = implicitRegionIdSupplier.get();
+      logger.debug(String.format("Using region id %s", regionId));
+      final Supplier<Set<String>> datacenterIdsSupplier = regionIdToZoneIdsSupplier.get().get(regionId);
+      this.datacenterIds = datacenterIdsSupplier.get();
+      logger.debug(String.format("Using datacenterIds %s", datacenterIds));
    }
 
    @Override
    public HttpRequest filter(HttpRequest request) throws HttpException {
-      Set<String> datacenterIds = datacenterIdsSupplier.get();
       if (datacenterIds != null && !datacenterIds.isEmpty()) {
-         return request.toBuilder().addQueryParam("datacenterId", datacenterIds).build();
+         return request.toBuilder().addQueryParam(queryParam, datacenterIds).build();
       } else {
          return request;
       }
```

Comment:
Having this as a filter might be problematic in the future, as the filter works with the "default region". When users start using the ComputeService and configure the TempalteOptions to deploy to a concrete region, there would be no way to call the InfraApi (or any API using this filter) and make it filter by the configured region.
It would be better to remove this filter and add the datacenterId as method parameters in the API.

Reply:
@nacx as agreed I have removed both of the HTTP filters.